en/lessons/libraries/bypass.md

```diff
@@ -0,0 +1,274 @@
+
+In this lesson we're going to explore how [bypass](https://github.com/PSPDFKit-labs/bypass) can help us quickly and easily handle these requests in our tests.
+
+{% include toc.html %}
+
+## What is Bypass?
+
+[Bypass](https://github.com/PSPDFKit-labs/bypass) is described as "a quick way to create a custom plug that can be put in place instead of an actual HTTP server to return prebaked responses to client requests."
+
+What does that mean?
+Under-the-hood Bypass is an OTP application that masquerades as an external server listening for and responding to requests.
+By responding with pre-defined responses you can test any number of possibilities such as services being unavailable, unexpected server errors, and any expected errors you might encounter.
+
+## Using Bypass
+
+To better illustrate the features of Bypass we'll be building a simple utility application to ping a list of domains and ensure they're online.
+To do this we'll create new supervisor project and a GenServer to check the domains on a configurable interval.
+By leveraging Bypass in our tests we'll be able to verify our application will work in many different outcomes.
+
+_Note_: If you wish to skip ahead to the final code, head over to the Elixir School repo [Clinic](https://github.com/elixirschool/clinic) and have a look.
+
+By this point we should be comfortable creating new Mix projects and adding our dependencies so we'll focus instead of the pieces of code we'll be testing.
+Let's start by creating a new module that will handle making the requests to our domains.
+
+Using [HTTPoison](https://github.com/edgurgel/httpoison) let's create a function, `ping/1`, that takes a URL and returns `{:ok, body}` for HTTP 200 requests and `{:error, reason}` for all others:
+
+```elixir
+defmodule Clinic.HealthCheck do
+  def ping(urls) when is_list(urls), do: Enum.map(urls, &ping/1)
+
+  def ping(url) do
+    url
+    |> HTTPoison.get()
+    |> response()
+  end
+
+  defp response({:ok, %{status_code: 200, body: body}}), do: {:ok, body}
+  defp response({:ok, %{status_code: status_code}}), do: {:error, "HTTP Status #{status_code}"}
+  defp response({:error, %{reason: reason}), do: {:error, reason}
+end
+
+```
+
+You'll notice we are _not_ making a GenServer and that's for good reason:
+By separating our functionality (and concerns) from the GenServer, we are able to test our code without the added hurdle of concurrency.
+
+With our code in place we need to start on our tests.
+Before we can use Bypass we'll need to ensure it's running.
+To do that, let's update `test/test_helper.exs` look like this:
+
+```elixir
+ExUnit.start()
+Application.ensure_all_started(:bypass)
+```
+
+Now that we know Bypass will be running during our tests let's head over to `test/clinic/health_check_test.exs` and finish our setup.
+To prepare Bypass for accepting requests we need to open the connect with `Bypass.open/1`, this can be done in our test setup callback:
```

Comment:
You are using Bypass.open/0. Maybe say this as well?
`Bypass.open/0` opens a connection with a default port that you can retrieve from the bypass struct using bypass.port. If you want to specify the listening port, you can use `Bypass.open/1`, like this: `Bypass.open(port: 1234)`

Reply:
I didnt realize this - thats super handy! CC @bmrsny cuz you could so use that in some of our old controller tests that grabbed it from env because it wasnt deterministic

Reply:
Good suggestion @gemantzu!  I use `Bypass.open(port: ...)` in the later section but a note here would be good 👍